File: PAT_Basic_Level_by_java/1062 最简分数/1062 最简分数.java
/**
 * 1062 最简分数
 * 
 * 题目描述：
 * 给定两个正分数 N1/M1 和 N2/M2，以及一个正整数 K，
 * 要求找出所有分母为 K 且值在 (N1/M1, N2/M2) 区间内的最简分数。
 * 最简分数是指分子和分母互质（最大公约数为 1）的分数。
 * 输出按分子从小到大排序。
 * 
 * 实现原理：
 * 解题思路：
 * 1. 先将两个分数统一为分母 K 的形式：比较 N1/M1 和 N2/M2 的大小，确保区间起始小于结束。
 * 2. 计算起始分子 start = N1 * K / M1 + 1（取上整），结束分子 end = N2 * K / M2（若整除则减一取下整）。
 * 3. 遍历 [start, end] 范围内的每个分子 i，用辗转相除法判断 i 与 K 是否互质（gcd(i, K) == 1）。
 * 4. 将所有满足条件的分子按分母 K 格式化为分数并输出。
 * 
 * 算法核心逻辑：通分确定分子范围 → 遍历判断互质 → 输出最简分数。
 * 
 * 时间复杂度：O(K log K)，空间复杂度：O(K)。
 */
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] parts = br.readLine().trim().split("\\s+");

        // 解析第一个分数 N1/M1
        String[] frac1 = parts[0].split("/");
        int n1 = Integer.parseInt(frac1[0]);
        int m1 = Integer.parseInt(frac1[1]);

        // 解析第二个分数 N2/M2
        String[] frac2 = parts[1].split("/");
        int n2 = Integer.parseInt(frac2[0]);
        int m2 = Integer.parseInt(frac2[1]);

        int K = Integer.parseInt(parts[2]); // 指定的分母

        // 确保第一个分数小于第二个分数，否则交换
        if (n1 * m2 > n2 * m1) {
            int tempN = n1, tempM = m1;
            n1 = n2;
            m1 = m2;
            n2 = tempN;
            m2 = tempM;
        }

        List<Integer> result = new ArrayList<>();

        // 计算起始分子：大于 N1/M1 的最小整数（向上取整）
        int start = n1 * K / m1 + 1;
        if (n1 * K % m1 == 0) {
            start = n1 * K / m1 + 1; // 恰好整除时，取下一个整数
        }

        // 计算结束分子：小于 N2/M2 的最大整数（向下取整）
        int end = n2 * K / m2;
        if (n2 * K % m2 == 0) {
            end--; // 恰好整除时，要严格小于 N2/M2
        }

        // 遍历所有候选分子，检查是否与 K 互质
        for (int i = start; i <= end; i++) {
            if (gcd(i, K) == 1) {
                result.add(i);
            }
        }

        // 格式化输出
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < result.size(); i++) {
            if (i > 0) {
                sb.append(" ");
            }
            sb.append(result.get(i)).append("/").append(K);
        }
        System.out.println(sb.toString());
    }

    // 辗转相除法求最大公约数
    private static int gcd(int a, int b) {
        while (b != 0) {
            int temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }
}

File: PAT_Basic_Level_by_c/1062 最简分数/1062 最简分数.c
/*
 * 题目：1062 最简分数
 * 实现原理：
 *   找出分母为k且值在两个分数之间的最简分数。
 *   算法步骤：
 *   1. 将两个分数转换为double值，确保d1 < d2
 *   2. 遍历i从1到k-1，检查i/k是否为最简分数（gcd(i,k)==1）
 *   3. 检查i/k是否在[d1, d2]区间内
 *   4. 输出满足条件的分数（用空格分隔）
 *   注意：最简分数即分子和分母的最大公约数为1。
 */
#include <stdio.h>

int gcd(int a, int b) {  // 辗转相除法求最大公约数
    return b == 0 ? a : gcd(b, a % b);
}

int main() {
    int n1, m1, n2, m2, k;  // 两个分数 n1/m1、n2/m2 及目标分母 k
    scanf("%d/%d %d/%d %d", &n1, &m1, &n2, &m2, &k);
    
    double d1 = (double)n1 / m1;  // 转成 double 便于比较大小
    double d2 = (double)n2 / m2;
    
    if (d1 > d2) {  // 保证 d1 <= d2
        double temp = d1;
        d1 = d2;
        d2 = temp;
    }
    
    int first = 1;  // 控制输出格式：第一个分数前不加空格
    for (int i = 1; i < k; i++) {  // 枚举可能的分子
        if (gcd(i, k) == 1) {  // 分子分母互质才是最简分数
            double d = (double)i / k;
            if (d > d1 && d < d2) {  // 值严格落在两个给定分数之间
                if (first) {
                    printf("%d/%d", i, k);
                    first = 0;
                } else {
                    printf(" %d/%d", i, k);
                }
            }
        }
    }
    printf("\n");
    
    return 0;
}